### PR TITLE
:bug: Address React 18 effect bug for useMountedRef

### LINKED ISSUE #2500: [@shopify/react-form] `form.submitErrors` is not populated  when returning `{status: "fail"}`
## Overview

In a very simple example `submitErrors` is not populated.

## Consuming repo

> What repo were you working in when this issue occurred?

Made a reproducable example. https://stackblitz.com/edit/vitejs-vite-qctwdk?file=src/App.tsx

---

## Checklist

- [x] Please delete the labels section before submitting your issue
- [x] I have described this issue in a way that is actionable (if possible)

### PR DESCRIPTION
**Pull request by QuintonC**
## Description

Fixes #2500

This addresses a bug in React 18 StrictMode environments where effects are run twice on initialization. Previously, the `useMountedRef` bug was initialized via `const mounted = useRef(true)`, but since the effect runs twice in development (including cleanups), the `ref` would be set to `false` and never re-initialized as `true`.

You can read more about the way effects are run in strict mode in the following places:
- [Updates to Strict Mode](https://react.dev/blog/2022/03/08/react-18-upgrade-guide#updates-to-strict-mode)
- [How to handle the Effect firing twice in development?](https://react.dev/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development)

I'm uncertain where else this bug might be present (via the callbacks being run twice in React 18 w/ Strict Mode), but I do know that this is one instance of it that was causing bugs within `react-form` (see the linked issue that this resolves). I've tested this inside of a 1P app and can provide folks on the team with a demo/spin link for testing if desired.

### CODE REVIEW

File: packages/react-hooks/src/hooks/mounted-ref.ts [hunk 1]
```diff
@@ -3,9 +3,11 @@ import {useRef} from 'react';
 import {useIsomorphicLayoutEffect} from './isomorphic-layout-effect';
 
 export function useMountedRef() {
-  const mounted = useRef(true);
+  const mounted = useRef(false);
```
Comment (BPScott) [inline packages/react-hooks/src/hooks/mounted-ref.ts:6]: What's the thinking behind changing this default value.

My initial thoughts would be I think we want this to retain true as its default value. This would be the value that is used between the first render when useIsomorphicLayoutEffect is called.

On a server, useIsomorphicLayoutEffect uses a plain useEffect (as useLayoutEffect errors on the  server), which runs after initial rendering so there would be a moment when this would return false, which would be different to current behaviour.

**Comment by QuintonC** [comment]
Updated a test instance which is using React 18 with Strict Mode enabled and getting the expected results within `react-form`.

**Review by BPScott — APPROVED** [review]
Great stuff! Thanks for spotting this and digging into the fix. "Only affects React 18 in StrictMode" is a fun edge-case.

Go hit squash and merge when you're happy.

**Comment by BPScott** [comment]
I took the liberty of adding a test case that only fails in React18.

**Comment by QuintonC** [comment]
> I took the liberty of adding a test case that only fails in React18.

Nice! 

Thank you for that and merging. My evening filled up yesterday and this got put to the side.
